### PR TITLE
202309 Notebook on scalar time functions

### PR DESCRIPTION
**Pull request by petermarshallio**

### CODE REVIEW

File: notebooks/03-query/07-functions-datetime.ipynb [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+   "id": "bbdbf6ad-ca7b-40f5-8ca3-1070f4a3ee42",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true
+   },
+   "source": [
+    "## Prerequisites\n",
+    "\n",
+    "This tutorial works with Druid 27.0.0 or later.\n",
+    "\n",
+    "Launch this tutorial and all prerequisites using the `druid-jupyter` profile of the Docker Compose file for Jupyter-based Druid tutorials. For more information, see [Docker for Jupyter Notebook tutorials](https://druid.apache.org/docs/latest/tutorials/tutorial-jupyter-docker.html)."
```
Comment (sergioferragut) [inline notebooks/03-query/07-functions-datetime.ipynb:42]: "using `druid-jupyter` or `all-services` profile"

File: notebooks/03-query/07-functions-datetime.ipynb [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "472589e4-1026-4b3b-bb79-eedabb2b44c4",
+   "metadata": {},
+   "source": [
+    "### Load example data\n",
+    "\n",
+    "Once your Druid environment is up and running, ingest the sample data for this tutorial.\n",
+    "\n",
+    "Run the following cell to create a table called `example-koalas-fndatetime`. Notice {the use of X as a timestamp | only required columns are ingested | WHERE / expressions / GROUP BY are front-loaded | partitions on X period and clusters by Y}.\n",
```
Comment (sergioferragut) [inline notebooks/03-query/07-functions-datetime.ipynb:91]: seems like "Notice {the use of X as a timestamp | only required columns are ingested | WHERE / expressions / GROUP BY are front-loaded | partitions on X period and clusters by Y}." is a template and not the final text.

File: notebooks/03-query/07-functions-datetime.ipynb [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+   "metadata": {},
+   "source": [
+    "## Use functions to format the date and time\n",
+    "\n",
+    "Below are examples of:\n",
+    "\n",
+    "* Using `TIME_FORMAT` to apply a string pattern to a timestamp.\n",
+    "* Extracting elements of a timestamp using `TIME_EXTRACT` and `EXTRACT`.\n",
+    "\n",
+    "The `TIME_FORMAT` function applies a [JODA](https://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeFormat.html) picture to a datetime column.\n",
```
Comment (sergioferragut) [inline notebooks/03-query/07-functions-datetime.ipynb:458]: picture or pattern/template?

File: notebooks/03-query/07-functions-datetime.ipynb [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+    "  EXTRACT(HOUR FROM __time) AS \"start-hour\",\n",
+    "  EXTRACT(HOUR FROM MILLIS_TO_TIMESTAMP(\"ended_at\")) AS \"end-hour\",\n",
+    "  COUNT(*) AS \"sessions\"\n",
+    "FROM \"example-koalas-fndatetime\"\n",
+    "GROUP BY 1, 2\n",
+    "'''\n",
+    "\n",
+    "df = pd.DataFrame(sql_client.sql(sql))\n",
+    "df_group=df.groupby(['start-hour','end-hour']).sum().unstack()\n",
+    "df_group.plot.bar(stacked=\"true\", legend=None)"
```
Comment (sergioferragut) [inline notebooks/03-query/07-functions-datetime.ipynb:509]: why no legend?

File: notebooks/03-query/07-functions-datetime.ipynb [hunk 5]
```diff
@@ -0,0 +1,963 @@
...
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "625b2d58-42e0-4ae7-bfd9-6f1d8eccf2ad",
+   "metadata": {},
+   "source": [
+    "The `GROUP BY` function can be used at ingestion time (also known as rollup in JSON-based ingestion) to pre-calculate commonly used aggregations.\n",
```
Comment (sergioferragut) [inline notebooks/03-query/07-functions-datetime.ipynb:729]: Should we call it "JSON-based ingestion" or just Native Ingestion?

**Review by sergioferragut — CHANGES_REQUESTED** [review]

File: notebooks/03-query/07-functions-datetime.ipynb (hunk 2 above)
Comment (petermarshallio) [reply notebooks/03-query/07-functions-datetime.ipynb:91]: Oops thanks!

File: notebooks/03-query/07-functions-datetime.ipynb (hunk 3 above)
Comment (petermarshallio) [reply notebooks/03-query/07-functions-datetime.ipynb:458]: You're right – their docs even say "pattern".

File: notebooks/03-query/07-functions-datetime.ipynb (hunk 4 above)
Comment (petermarshallio) [reply notebooks/03-query/07-functions-datetime.ipynb:509]: I just felt it looked a bit cluttered when it generated...

File: notebooks/03-query/07-functions-datetime.ipynb (hunk 5 above)
Comment (petermarshallio) [reply notebooks/03-query/07-functions-datetime.ipynb:729]: Good spot. I checked the docs and it's "Classic batch" and "streaming" are separated, with "JSON-based" being used specifically to refer to batch ingestion. I've therefore just updated this notebook with "classic ingestion" throughout, and mentioned both batch and streaming.

**Pull request merged**